core/services/workflows/engine.go

```diff
@@ -134,204 +165,335 @@ func (e *Engine) registerTrigger(ctx context.Context) error {
+			if innerErr != nil {
+				return fmt.Errorf("failed to unregister from workflow: %+v", reg)
+			}
+
+			return nil
+		})
 		if err != nil {
 			return err
 		}
-		deregRequest := capabilities.CapabilityRequest{
-			Metadata: capabilities.RequestMetadata{
-				WorkflowID: mockedWorkflowID,
-			},
-			Inputs: triggerInputs,
-		}
-		return e.trigger.UnregisterTrigger(context.Background(), deregRequest)
+
+		return nil
 	})
 }
 
-func NewEngine(lggr logger.Logger, registry types.CapabilitiesRegistry) (engine *Engine, err error) {
-	yamlWorkflowSpec := `
-triggers:
-  - type: "on_mercury_report"
-    ref: report_data
-    config:
-      feedlist:
-        - "0x1111111111111111111100000000000000000000000000000000000000000000" # ETHUSD
-        - "0x2222222222222222222200000000000000000000000000000000000000000000" # LINKUSD
-        - "0x3333333333333333333300000000000000000000000000000000000000000000" # BTCUSD
-        
-consensus:
-  - type: "offchain_reporting"
-    ref: evm_median
-    inputs:
-      observations:
-        - $(report_data.outputs)
-    config:
-      aggregation_method: data_feeds_2_0
-      aggregation_config:
-        0x1111111111111111111100000000000000000000000000000000000000000000:
-          deviation: "0.001"
-          heartbeat: "30m"
-        0x2222222222222222222200000000000000000000000000000000000000000000:
-          deviation: "0.001"
-          heartbeat: "30m"
-        0x3333333333333333333300000000000000000000000000000000000000000000:
-          deviation: "0.001"
-          heartbeat: "30m"
-      encoder: EVM
-      encoder_config:
-        abi: "mercury_reports bytes[]"
-
-targets:
-  - type: write_polygon-testnet-mumbai
-    inputs:
-      report:
-        - $(evm_median.outputs.reports)
-    config:
-      address: "0x3F3554832c636721F1fD1822Ccca0354576741Ef"
-      params: [($inputs.report)]
-      abi: "receive(report bytes)"
-  - type: write_ethereum-testnet-sepolia
-    inputs:
-      report:
-        - $(evm_median.outputs.reports)
-    config:
-      address: "0x54e220867af6683aE6DcBF535B4f952cB5116510"
-      params: ["$(inputs.report)"]
-      abi: "receive(report bytes)"
-`
-
-	workflow, err := Parse(yamlWorkflowSpec)
+type Config struct {
+	Spec           string
+	Lggr           logger.Logger
+	Registry       types.CapabilitiesRegistry
+	MaxWorkerLimit int
+}
+
+const (
+	defaultWorkerLimit = 100
+)
+
+func NewEngine(cfg Config) (engine *Engine, err error) {
+	if cfg.MaxWorkerLimit == 0 {
+		cfg.MaxWorkerLimit = defaultWorkerLimit
+	}
+	// TODO: validation of the workflow spec
+	// We'll need to check, among other things:
+	// - that there are no node `ref` called `trigger` as this is reserved for any triggers
+	// - that there are no duplicate `ref`s
+	// - that the `ref` for any triggers is empty -- and filled in with `trigger`
+	// - etc.
+
+	workflow, err := Parse(cfg.Spec)
 	if err != nil {
 		return nil, err
 	}
+
+	// Instantiate semaphore to put a limit on the number of workers
```

Comment:
Is that an idiomatic way in Go to create a thread pool? ;) Maybe there is some well-known library that does that for us? Or maybe it'd be simpler to just count them and fail new executions when it reaches some crazy number? cc @jmank88 any suggestions?

Reply:
I think so... Do we need a well-known library to fill up a channel with some tokens that we pick up? It's ~3 lines of code.

Reply:
An alternative would be to just have the full worker pool spun up in `Start` but that obviously means we have no "flex" in the system, although note that the runtime cost of a bunch of paused goroutines is very low so maybe it's good enough 🤷 

I picked this approach because it has the benefit of allowing us to pick a high number and then we'll only use what we need at any given point in time.

Reply:
Makes sense, let's keep it then.

Reply:
This seems to be popular https://github.com/alitto/pond